src/EventSubscriber/ElasticApmRequestSubscriber.php

```diff
@@ -0,0 +1,238 @@
+    }
+
+    // Mark the request as processed.
+    $this->processedMasterRequest = TRUE;
+  }
+
+  /**
+   * End the transaction and send to PHP Agent whenever this event is triggered.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\FinishRequestEvent $event
+   *   The event to process.
+   */
+  public function onFinishRequest(FinishRequestEvent $event) {
+    // Don't process if Elastic APM is not configured.
+    if (!$this->elasticApm->isConfigured()) {
+      return;
+    }
+
+    // Don't process if we don't have a PHP Agent already initialized, meaning,
+    // no transaction is in process.
+    if (!$this->phpAgent) {
+      return;
+    }
+
+    // End the transaction.
+    try {
+      $this->phpAgent->stopTransaction($this->routeMatch->getRouteName());
+
+      // Send our transaction to Elastic.
+      $this->phpAgent->send();
+    }
+    catch (Exception $e) {
+      // Notify the user of the error.
+      $this->messenger->addError(t('An error occurred while trying to send the transaction to the Elastic APM server.'));
+
+      // Log the error to watchdog.
+      $error = Error::decodeException($e);
+      $this->logger->log($error['severity_level'], '%type: @message in %function (line %line of %file).', $error);
+    }
+  }
+
+  /**
+   * Create spans around the db queries that were run in this request.
+   *
+   * @return array
+   *   An array of spans that will be added to the transaction.
+   */
+  protected function constructDatabaseSpans() {
+    $spans = [];
+
+    // First, go through all queries that have been run for this request.
+    $connections = [];
+    foreach (Database::getAllConnectionInfo() as $key => $info) {
+      $database = Database::getConnection('default', $key);
+      $connections[$key] = $database->getLogger()->get('elastic_apm');
+    }
+
+    // Now, create a span for each query that was run.
+    foreach ($connections as $key => $queries) {
+      foreach ($queries as $query) {
+        $span = [];
+        // Add the necessary schema info for the APM server.
+        $span['name'] = $query['caller']['function'];
+        $span['type'] = 'db.mysql.query';
+        // This is the query start time relative to the transaction start.
+        $span['start'] = 0;
+        // Change duration time of query to milliseconds.
+        $span['duration'] = $query['time'] * 1000;
+        $span['context'] = [
+          'db' => [
+            'instance' => $key,
+            'statement' => $query['query'],
+            'type' => 'sql',
+          ],
+        ];
+        $span['stacktrace'] = [
+          [
+            'function' => $query['caller']['function'],
+            'abs_path' => $query['caller']['file'],
+            'filename' => substr($query['caller']['file'], strrpos($query['caller']['file'], '/') + 1),
+            'lineno' => $query['caller']['line'],
+            'vars' => $query['args'],
+          ],
+        ];
+
+        $spans[] = $span;
+      }
+    }
```

Comment:
**OPTIONAL CHANGE REQUEST**

*Readability/Code structure*

Could we pull creating the span for an individual query into a separate functions to avoid having functions of 30+ lines?